In [1]:
# Lab settings - please ingnore
options(repr.plot.width=7, repr.plot.height=4, repr.plot.res=250 ) # Make plots a resonable size

<div class="big_title">LAB 7: Flow control and Functions Explained</div>

BIO3782: Biologist's Toolkit (Dalhousie University)

----------------------------------------

# Setup of workspace

<span class="important"></span> Make sure the required files are in the working directory:

* Create a **folder**  on the "desktop" and name it <span class="file">Lab7</span>
* In RStudio, change the working directory to <span class="file">Desktop/Lab7</span>. Click here if you need a [refresher on the **working directory**](https://diego-ibarra.github.io/biol3782/week1/index.html#RStudio's-%22Working-Directory%22)
* In RStudio, create a new R script, name it <span class="file">lab7.r</span> and make sure you save it in <span class="file">Desktop\Lab7</span>. You will be writing and copy-pasting code to your <span class="file">lab7.r</span> file so that you can keep a record of all you did in this lab.


As in previous labs, we'll try simulate "real-life" coding, by using the tags below to indicate when to use RStudio's <span class="editor"></span> and when to use the <span class="console"></span>: 


<br>
<div class="use_editor"></div>
<br><br>

<div class="use_console"></div>
<br><br>

----------------------------------------

# Flow Control

So far we have written R scripts that are linear; that is, statments are executed in a sequential manner like the diagram below:

<img src="flow1.png" width="100px">

However, there are ways to execute statments using "bifurcations" and "go-back" directives that allow the programmer to control the flow of statement execution. Below we explore some of this ways of flow control.

##  IF statements

**IF statements**, also known as conditional statements or conditional expressions, allow programming languages to perform different computations or actions depending on whether a programmer-specified boolean condition evaluates to `TRUE` or `FALSE`.

<img src="if1.png" width="300px">

In the example shown in the diagram above, when R evaluates the **Condition**, if the result is `TRUE`, then the statement on the **left** is executed. On the other hand, if the result is `FALSE`, then the statement on the **right** is executed.

There are many types of **IF statements**. Below we review a few:

### IF

In the simplest type of **IF statement**, R will execute a statement (or set of statements) only if the evaluated condition is `TRUE` (as always). However, if the condition result is `FALSE` R will simply **do nothing**. Take a look at the diagram below:

<img src="if2.png" width="280px">


The syntax of this simplest case of IF stament is:
<br><br>

<div class="nomenclature">
if (condition is true) {<br>
&nbsp;&nbsp;perform action<br>
}<br>
</div> 

First let's do an example where the condition is `TRUE`:

<br>
<div class="use_editor"></div>

In [2]:
# Example of IF statement where condition is TRUE

x <- 12

if (x >= 10) {
  print("x is greater than or equal to 10")
}

x

[1] "x is greater than or equal to 10"


[1] 12

As you can see, the condition `(x >= 10)` was `TRUE`, thus the statement `print("x is greater than or equal to 10")` was executed. 

You probably evaluated the condition `(x >= 10)` in you head and know that the result is `TRUE`. However, if you have a more complex conditional, you can simply copy-paste the condition to into the <span class="console"></span> and let R do the evaluation:

<br>
<div class="use_console"></div>

In [3]:
(x >= 10)

[1] TRUE

Now let's try an example where the condition is `FALSE`:


<br>
<div class="use_editor"></div>

In [2]:
# Example of IF statement where condition is FALSE

x <- 5

if (x >= 10) {
  print("x is greater than or equal to 10")
}

x

[1] "x is greater than or equal to 10"


[1] 12

As you can see,  the condition `(x >= 10)` was `FALSE`, in therefore R simply **did nothing**!

###  IF ... ELSE

In the case where you want R to execute different statements depending on the result of the conditional, you can use **"IF ... ELSE"** block. Take a look how it works:


<img src="if3.png" width="300px">

The syntax of an **"IF ... ELSE"** block is:

<br><br>

<div class="nomenclature">
if (condition is true) {<br>
&nbsp;&nbsp;perform action<br>
} else {  # that is, if the condition is false,<br>
&nbsp;&nbsp;perform alternative action<br>
}
</div> 

<br><br>

Take a look at the example below:

<br>
<div class="use_editor"></div>

In [4]:
# Example of IF...ELSE statement where condition is FALSE

x <- 8

if (x >= 10) {
  print("x is greater than or equal to 10")
} else {
  print("x is less than 10")
}

[1] "x is less than 10"


As you can see, if the condition is `FALSE` **the ELSE statement is executed**... instead of simply "do nothing", like in the example above (i.e. IF example).

If the condition is `TRUE`, the behaivor is the same as the with the "simple IF statement". 

<br>
<div class="use_editor"></div>

In [4]:
# Example of IF...ELSE statement where condition is TRUE

x <- 12

if (x >= 10) {
  print("x is greater than or equal to 10")
} else {
  print("x is less than 10")
}

[1] "x is greater than or equal to 10"


### IF ... ELSE IF ... ELSE

If you want to "daisy chain" multiple conditions, you can use one (or more) **`else if`** statements. Take a look at the diagram below:

<img src="if4.png" width="450px">

You can see that if the **first condition** is `FALSE`, then the flow is diverted to the **second condition**. If that result is `FALSE`, then the flow will be diverted to the **third condition**, ans so on. Note that you can have many **`else if `** statements chained together. This type of block **ALWAYS** must start with an **`if`** statement. However, it may or may not finish with an **`else`** statement depending on whether you want the last action to be "do nothing".

The syntax of an **"IF ... ELSE IF ... ELSE"** block is:

<br><br>

<div class="nomenclature">
if (condition is true) {<br>
&nbsp;&nbsp;perform action<br>
} else if (condition is true) {  # that is, if the condition is false,<br>
&nbsp;&nbsp;perform alternative action<br>
} else {  # that is, if the condition is false,<br>
&nbsp;&nbsp;perform alternative action<br>
}
</div> 

<br>

Take a look at the example below:

<br>
<div class="use_editor"></div>

In [5]:
# Example of IF...ELSE IF...ELSE statement where second condition is TRUE
x <- 8

if (x >= 10) {
  print("x is greater than or equal to 10")
} else if (x > 5) {
  print("x is greater than 5, but less than 10")
} else {
  print("x is less than 5")
}

[1] "x is greater than 5, but less than 10"


Iin this case `(x >= 10)` was `FALSE`, thus the second condition `(x > 5)`was evaluated and returned `TRUE`.

Let's try a case where all conditions in the block are `FALSE`:


<br>
<div class="use_editor"></div>

In [6]:
x <- 2

if (x >= 10) {
  print("x is greater than or equal to 10")
} else if (x > 5) {
  print("x is greater than 5, but less than 10")
} else {
  print("x is less than 5")
}

[1] "x is less than 5"


Finally, for completeness, let's try an example where the first condion is `TRUE`:


<br>
<div class="use_editor"></div>

In [7]:
x <- 12

if (x >= 10) {
  print("x is greater than or equal to 10")
} else if (x > 5) {
  print("x is greater than 5, but less than 10")
} else {
  print("x is less than 5")
}

[1] "x is greater than or equal to 10"


## FOR Loops

If you want to iterate over a set of values, when the order of iteration is important, and perform the same operation on each, a `for(`) loop will do the job. We saw `for()` loops in the shell lessons earlier. This is the most flexible of looping operations, but therefore also the hardest to use correctly. In general, the advice of many R users would be to learn about `for()` loops, but to avoid using `for()` loops unless the order of iteration is important: i.e. the calculation at each iteration depends on the results of previous iterations. If the order of iteration is not important, then you should learn about vectorized alternatives, such as the `purr` package, as they pay off in computational efficiency.


<img src="for_loop.png" width="300px">


The basic structure of a `for()` loop is:


<div class="nomenclature">
for (iterator in set of values) {<br>
&nbsp;&nbsp;do a thing<br>
}<br>
</div> 

For example:

In [8]:
for (i in 1:10) {
  print(i)
}

[1] 1
[1] 2
[1] 3
[1] 4
[1] 5
[1] 6
[1] 7
[1] 8
[1] 9
[1] 10


The `1:10` bit creates a vector on the fly; you can iterate over any other vector as well.

We can use a `for()` loop nested within another `for()` loop to iterate over two things at once.

In [1]:
for (i in 1:5) {
  for (j in c('a', 'b', 'c', 'd', 'e')) {
    print(paste(i,j))
  }
}

[1] "1 a"
[1] "1 b"
[1] "1 c"
[1] "1 d"
[1] "1 e"
[1] "2 a"
[1] "2 b"
[1] "2 c"
[1] "2 d"
[1] "2 e"
[1] "3 a"
[1] "3 b"
[1] "3 c"
[1] "3 d"
[1] "3 e"
[1] "4 a"
[1] "4 b"
[1] "4 c"
[1] "4 d"
[1] "4 e"
[1] "5 a"
[1] "5 b"
[1] "5 c"
[1] "5 d"
[1] "5 e"


In [2]:
output_vector <- c()
for (i in 1:5) {
  for (j in c('a', 'b', 'c', 'd', 'e')) {
    temp_output <- paste(i, j)
    output_vector <- c(output_vector, temp_output)
  }
}
output_vector

[1] "1 a" "1 b" "1 c" "1 d" "1 e" "2 a" "2 b" "2 c" "2 d" "2 e" "3 a" "3 b"
[13] "3 c" "3 d" "3 e" "4 a" "4 b" "4 c" "4 d" "4 e" "5 a" "5 b" "5 c" "5 d"
[25] "5 e"

In [13]:
# Start the clock!
ptm <- proc.time()

output_vector <- c()
for (i in 1:10000) {
  for (j in c('a', 'b', 'c', 'd', 'e')) {
    temp_output <- paste(i, j)
    output_vector <- c(output_vector, temp_output)
  }
}

str(output_vector)

# Stop the clock
proc.time() - ptm

 chr [1:50000] "1 a" "1 b" "1 c" "1 d" "1 e" "2 a" "2 b" "2 c" "2 d" "2 e" ...


   user  system elapsed 
  11.99    0.10   12.07 

In [16]:
# Start the clock!
ptm <- proc.time()

output_matrix <- matrix(nrow=10000, ncol=5)
j_vector <- c('a', 'b', 'c', 'd', 'e')
for (i in 1:10000) {
  for (j in 1:5) {
    temp_output <- paste(i, j_vector[j])
    output_matrix[i, j] <- temp_output
  }
}
output_vector2 <- as.vector(output_matrix)

str(output_vector)

# Stop the clock
proc.time() - ptm


 chr [1:50000] "1 a" "1 b" "1 c" "1 d" "1 e" "2 a" "2 b" "2 c" "2 d" "2 e" ...


   user  system elapsed 
   0.28    0.00    0.28 

## WHILE Loops

Sometimes you may need needing to repeat an operation as long as a certain condition is met. You can do this with a `while()` loop.

<div class="nomenclature">
while(this condition is true){<br>
&nbsp;&nbsp;do a thing<br>
}<br>
</div>

In the example below, the `while()` loop will run while the value of `x` is greater than zero. The value of `x` is re-calculated in every iteration of the `while()` loop (also the value of `x` is printed to screen):

In [19]:
x <- 5

while(x > 0) {
    print(x)
    x <- x-1
}

[1] 5
[1] 4
[1] 3
[1] 2


# Functions


<div class="nomenclature">
functionName &lt;- function(arguments) {<br>
&nbsp;&nbsp;do a thing<br>
&nbsp;&nbsp;return(output)<br>
}<br>
</div>

# Real-life example: NPZ model

Here we will built a numerical **N**utrients-**P**hytoplankton-**Z**ooplankton ecosystem model (i.e. an NPZ model) with the dynamics show in the diagram below. The "currency" of the model is nitrogen (i.e. $mmol N m^{-3}$) and the state variables are Nutrients ($N$), Phytoplankton ($P$) and Zooplankton ($Z$).

<img src="NPZmodel.png" width="500px">

In numerical models the programmer has to specify two things:

1. **Mathematical equations** discribing ecosystem processes like photosynthesis, mortality, grazing, etc., and
2. **Initial conditions** which represent the state of the ecosystem (i.e. concentrations of Phytoplankton, Zooplankton and Nutrients) at the beginning of the simulation.

The essence of these numerical models is a **FOR LOOP** where the **Mathematical equations** are used to calculate the **future** state of the model (i.e. the concentrations of Phytoplankton, Zooplankton and Nutrients) based on the **present** state. For every iteration of the **FOR LOOP**, the model estimates the state of the model at very small step in the **future** (i.e. only a few seconds into the future). In a a typical model run, the **FOR LOOP** can go around **hundreds of thousands of times**, effectively modelling years of ecosystem dynamics.

Let's take as an example the concentration of Phytoplankton, $P$, where the **present** concentration is $P_t$ and the **future** concentration is $P_{t+1}$. 

* In the first iteration of the **FOR LOOP**, the value of the **present** concentration $P_t$ is used to calculate the **future** concentration, $P_{t+1}$
* At the beginning of the second iteration of the **FOR LOOP**, the value of **present** concentration $P_t$ is updated from the previous calculation (i.e. New $P_t$ = the old $P_{t+1}$). Then, a new $P_{t+1}$ is calculated, and so on...
* Note that using the previous $P_t$ to calculate the new $P_{t+1}$ works fine in all iterations **except the first one!**. In the first iteration the value of $P_t$ is "initiallized" with an **"initial condition"** value specified by the programmer. 


## Generic equations

I teach a whole course on Ecosystem Modelling, where students learn how to design equations to program the model to do what they what it to do. However, in the interest of time, here I will simply give you the equations. As a programmer, your job is to get the equations supplied by the modeller and turn them into R code.

Using the diagram above as a reference, the following generic equations describe the dynamics among the Nutrients ($N$), Phytoplankton ($P$) and Zooplankton ($Z$) state variables:


(1) $$ \frac{\partial P}{\partial t} = photosynthesis - grazing - mortalityP $$

(2) $$  \frac{\partial N}{\partial t} = mortalityP + mortalityZ - photosynthesis $$

(3) $$  \frac{\partial Z}{\partial t} = grazing - mortalityZ $$


## Continuous equations

Using the "generic equations" above, we can define actual partial differential equations to describe the different ecosystem processes and flows of biomass among the different state variables (see table below for symbols and units):


(4) $$ \frac{\partial P}{\partial t} = \mu L^N P - \gamma L^P Z - m_P P$$

(5) $$  \frac{\partial N}{\partial t} = m_P P + m_ZZ - \mu L^N P $$

(6) $$  \frac{\partial Z}{\partial t} = \gamma L^P Z -m_ZZ $$

(7) $$L^N = \frac{N}{K+N}$$

(8) $$L^P = 1-e^{-\Lambda P}$$
<br><br>

Note that equations 1, 2 and 3 are equivalent to equations 4, 5, and 6, respectively. We just replaced the "process terms" with the actual mathematical terms representing those processes. 


Table of symbols and units:

Symbol &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; |  Units &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; |  Description
:---: | :--- | :---
$P$ | $mmol N m^{-3}$ | phytonplankton concentration
$N$ | $mmol N m^{-3}$ | nutrients concentration
$Z$ | $mmol N m^{-3}$ | zooplankton concentration
$L^N$ | dimensionless | Limitation due to (low) nutrients on phytoplankton (see Eq. 7)
$L^P$ | dimensionless | Limitation due to (low) phytoplankton on zooplankton (see Eq. 8)
$K$ | $mmol N m^{-3}$ | Half-saturation for nutrient absoption by phytoplankton
$t$ | $days$ | time
$\Lambda$ | $mmol N^{-1} m^3$ | Initial slope of ingestion saturation of zooplankton
$m_P$ | $d^{-1}$ | mortality rate of phytoplankton
$m_Z$ | $d^{-1}$ | mortality rate of zooplankton
$\mu$ | $d^{-1}$ | growth rate of phytoplankto
$\gamma$ | $d^{-1}$ | growth rate of zooplankton


## Discrete equations

The model equations above (i.e. Equations 4 to 8) need to be discretized to convert them into a form that can be processed by computers. We won't talk much about how the discretization process works. We only need to understand that, with some mathematical assumtions, Equations 4 to 8 are converted into Equations 9 to 13 below: 

(9) $$ P_{t+1} = P_t + \left\{ \left[ \mu L^N_t P_t - \gamma L^P_t Z_t - m_P P_t \right] \Delta t \right\} $$

(10) $$  N_{t+1} = N_t + \left\{ \left[ m_P P_t + m_Z Z_t - \mu L^N_t P_t \right] \Delta t \right\} $$

(11) $$  Z_{t+1} = Z_t +  \left\{ \left[ \gamma L^P_t Z_t -m_ZZ_t \right] \Delta t \right\} $$

(12) $$L^N_t = \frac{N_t}{K+N_t}$$

(13) $$L^P_t = 1-e^{-\Lambda P_t}$$

<br>

As you can see, this "discrete equations" are what we need to calculate the **future** using the **present**. For example, take a look at Equation 9, where we estimate the value of of **future** value of Phytoplankton, $P_{t+1}$, using only the the values of the present state (i.e. $P_t$, $Z_t$ and $N_t$).

## Transferring equations to computer code

In RStudio, create a new R script, name it <span class="file">NPZmodel.r</span> and make sure you save it in <span class="file">Desktop\Lab7</span>. Here on you will be required to copy-paste bits of code into your <span class="file">NPZmodel.r</span> file. We will simulate the normal process that someone may follow in the development of a new model from scratch.

Roughly speaking, the steps that we need to follow are the following:

1. Create empty vectors for each variable
2. Create the FOR LOOP where we will include Equations 9 to 13
3. Graph the results

There may be many intermediate steps in between the steps above, however we can figure those out as we go.


### Creating the "time" vector

The first step is to create a `time` vector. This will be a long column that starts at `0` and ends in the number of days that we want our model to run for (e.g. 365 days). Note that length of the vector will depend on the model's **"time-step"**, which is how much forward in time the model will advance in ever iteration of the FOR LOOP. Remember that for the model to work properly the time-step needs to be very small, in the order of a few seconds. 

Quickly Google-ing "Create vector of equally spaced numbers in R" revelas that we need to use the `seq()` and a quick look at the help file of the function (i.e. `?seq`) shows that we need 3 input arguments:

1. `from =`, which is the starting point of the vector, in our case: `0`
2. `to =`, which is the ending point of the vector. In our case, for example, we could make this `365` if we want to run the model for a year
3. `length =`, which the number of steps between the starting and ending points. We need to calculate this using the "time-step"


Type the following in your <span class="file">NPZmodel.r</span> file, and click <span class="source"></span>:

<br>
<div class="use_editor"></div>

In [6]:
# Framework
days <- 365 * 2 # number of days of experiment run (units: days)
dt   <- 0.01    # time-step  (units: days)

# Chores (calculate number of steps, create time vector, create zero vectors, ...)
NoSTEPS <- as.integer(days / dt) # Calculates the number of steps by dividing days by dt and rounding down
time <- seq(from = 0, to = days, length = NoSTEPS) # Create time vector of equally spaced numbers from zero to "days"

As you can see, in the first couple of line we define `days` and `dt`. The modeller (i.e. you) can change these parameters at any time. For example, if you want to run the model for 10 years just change to `days <- 365 * 10`... or if you want to make the "time-step" smaller, simply change to `dt   <- 0.001`.

It is a good practice to **keep at the top of your script all the parameters that are ment to be "adjusted" by the modeller**. These first few lines of code are like "knobs" that the modeller can adjust before running the model.

Note that we needed to do `NoSTEPS <- as.integer(days / dt)` to calculate the number of steps in the run, which was then used as the `length =` argument when making the `time` vector.

As you can see, the starting and focal point was to make the `time` vector via `time <- seq(from = 0, to = days, length = NoSTEPS)`. The rest ofthe code was just to calculate all the arguments needed by the function `seq()`.

Let's take a look at our newly created `time` vector:

<br>
<div class="use_console"></div>

In [14]:
str(time)

head(time)

tail(time)

 num [1:73000] 0 0.01 0.02 0.03 0.04 ...


[1] 0.00000000 0.01000014 0.02000027 0.03000041 0.04000055 0.05000068

[1] 729.95 729.96 729.97 729.98 729.99 730.00

<div class="Q"><br>

How many columns are in the <b>time</b> vector?
    
ANSWER: 1

<br><br><div class="qhelp"></div></div>

<div class="Q"><br>

How many rows are in the <b>time</b> vector?
    
ANSWER: 73000

<br><br><div class="qhelp"></div></div>

<div class="Q"><br>

What is the first number in the <b>time</b> vector?
    
ANSWER: 0

<br><br><div class="qhelp"></div></div>

<div class="Q"><br>

What is the last number in the <b>time</b> vector?
    
ANSWER: 730

<br><br><div class="qhelp"></div></div>

<div class="Q"><br>

What are the units of the <b>time</b> vector?
    
ANSWER: days

<br><br><div class="qhelp"></div></div>

### Creating empty vectors

The second step is to create an empty vector for every model variable (i.e. `N`, `P`, `Z`, `time` and others). 

<span class="important"></span> You should remember from the FOR LOOPS section above, that it is very important to pre-make empty vectors to avoid the loop to **drastically slow down!!**. During the FOR LOOP, the values of the empty vectors will be later filled out. 

The function to create empty vectors is `numeric()`. If you did not know this, you could have Googled "how to create an empty vector in R" as a starting point.

A quick `?numeric` to read the functions help file reveals that we just need one argument: length of vector. This is great because we already calculated this (i.e. `NoSTEPS`).

We need one empty vector for each of the equations 9 to 13. Go ahead and copy-paste the following at the end of the `# Chores` section of your code in your <span class="file">NPZmodel.r</span> file, and click <span class="source"></span>:

In [15]:
P <- numeric(NoSTEPS)      # Make Phytoplankton "empty" vector by making a vector array of zeros (size: NoSTEPS rows by ONE column) (units: mmolN m^-3) 
N <- numeric(NoSTEPS)      # Make Nutrients "empty" vector (same process as above) (units: mmolN m^-3)
Z <- numeric(NoSTEPS)      # Make Zooplankton "empty" vector (same process as above) (units: mmolN m^-3)
L_N <- numeric(NoSTEPS)    # Make "Limitation of N on Phytoplankton" empty vector (same process as above) (units: dimensionless)
L_P <- numeric(NoSTEPS)    # Make "Limitation of P on Zooplankton" empty vector (same process as above) (units: dimensionless)

Let's take a look at one our newly created empty vectors:

In [17]:
P

[1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
   [37] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
   [73] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [109] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [145] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [181] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [217] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [253] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [289] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [325] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [361] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [397] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [433] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [469] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [505] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [541] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [577] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [613] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [649] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [685] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [721] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [757] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [793] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [829] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [865] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [901] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [937] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  [973] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1009] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1045] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1081] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1117] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1153] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1189] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1225] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1261] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1297] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1333] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1369] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1405] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1441] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1477] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1513] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1549] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1585] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1621] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1657] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1693] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1729] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [1765] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [18

As you can see, `P` is just a bunch of zeros.

<div class="Q"><br>

How many columns are in the <b>N</b> vector?
    
ANSWER: 1

<br><br><div class="qhelp"></div></div>

<div class="Q"><br>

How many rows are in the <b>Z</b> vector?
    
ANSWER: 73000

<br><br><div class="qhelp"></div></div>

<div class="Q"><br>

What is the first number in the <b>N</b> vector?
    
ANSWER: 0

<br><br><div class="qhelp"></div></div>

<div class="Q"><br>

What is the last number in the <b>P</b> vector?
    
ANSWER: 730

<br><br><div class="qhelp"></div></div>

<div class="Q"><br>

What are the units of the <b>P</b> vector?
    
ANSWER: days

<br><br><div class="qhelp"></div></div>

In [16]:
str(P)

head(P)

tail(P)

 num [1:73000] 0 0 0 0 0 0 0 0 0 0 ...


[1] 0 0 0 0 0 0

[1] 0 0 0 0 0 0

<div class="big_title">This is the end of lab</div>

*******************
*******************

Code below is for formatting of this lab. Do not alter!

In [64]:
cssFile <- '../css/custom.css'
IRdisplay::display_html(readChar(cssFile, file.info(cssFile)$size))

IRdisplay::display_html("<style>body {counter-reset: question_num;}.Q::before {counter-increment: question_num;
    content: 'QUESTION ' counter(question_num) '\\A'; white-space: pre; }</style>")